src/components/Editor/CitationManager/index.js

```diff
@@ -0,0 +1,333 @@
+   * @param {String} msg message of error
+   * @param  {...any} params other parameters
+   */
+  constructor(msg, ...params) {
+    super(msg, ...params)
+
+    // Maintains stack trace (V8 only)
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, CitationOpError)
+    }
+
+    this.name = 'CitationOpError'
+  }
+}
+
+class CitationNode {
+  /**
+   * constructor of CitationNode
+   * @param {Node} node
+   */
+  constructor(node) {
+    this.node = node
+  }
+
+  /**
+   * get id of the editor that owns this node
+   */
+  get editorId() {
+    return this.node.$props.editor.view.dom.id
+  }
+
+  /**
+   * get pos of the node in the editor
+   */
+  get pos() {
+    return this.node.getPos()
+  }
+}
+
+class CitationInfo {
+  /**
+   * constructor
+   * @param {String} title : title of citation
+   * @param {String} url : url of citation
+   */
+  constructor(title, url) {
+    new URL(url)
+
+    if (title.length === 0) {
+      throw new CitationOpError('Citation.info: title should not be empty')
+    }
+
+    this.info = { title, url, index: -1 }
+    this.nodes = []
+  }
+
+  /**
+   * connect node with citation
+   *
+   * @param {CitationNode} node node to be connected
+   */
+  addNode(node) {
+    const editorId = node.$props.editor.view.dom.id
+    const citationNode = new CitationNode(node, editorId)
+
+    this.nodes.push(citationNode)
+    this.sortNodesArray()
+  }
+
+  /**
+   *
+   * @param {CitationNode} node node to be removed
+   * @returns Boolean true if list of node is empty after removing
+   */
+  removeNode(node) {
+    this.nodes.splice(this.nodes.indexOf(node), 1)
+
+    if (this.nodes.length === 0) return true
+    return false
+  }
+
+  /**
+   * delete all node in next tick
+   */
+  removeAllNode() {
+    for (const node of this.nodes) {
+      node.node.$nextTick(() => node.node.deleteNode())
+    }
+  }
+
+  /**
+   * update citation info
+   * @param {String} title : new title
+   * @param {String} url : new url
+   * @throws {TypeError} : if a invalid url passed
+   * @throws {CitationOpError} : if an empty title passed
+   */
+  updateInfo({ title, url }) {
+    new URL(url)
+
+    if (title.length === 0) {
+      throw new CitationOpError('Citation.info: title should not be empty')
+    }
+
+    this.info = Object.assign({}, this.info, { title, url })
+
+    const ids = this.nodes.flatMap((n) => n.editorId)
+    for (const id of ids) {
+      this.nodes[0].node.$nextTick(() => {
+        const content = store.state.post.currentEditors[id].getJSON()
+        store.commit('post/UPDATE_BLOCK_CONTENT', {
+          id,
+          content
+        })
+      })
+    }
+  }
+
+  /**
+   * update index of this citation
+   * @param {Number} index new index of node to be updated
+   */
+  updateIndex(index) {
+    this.info = Object.assign({}, this.info, { index })
+  }
+
+  /**
+   * sort list of nodes from top to bottom
+   */
+  async sortNodesArray() {
```

Comment:
No async action found in scope. Do we need async?

Reply:
The reason that using `async` here is to make `sortNodesArray()` asynchronous to avoid modal hang after `submit` button clicked.
"Async functions may also be defined as expressions", [MDN document](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function).